### DISCUSSION TITLE
How often and how many times does `commands.Bot` attempt to connect to a channel?

### DISCUSSION THREAD

**Discussion by anakimluke** [TwitchIO/TwitchIO#400] [Questions] [answered]
I had a bot in a working stage online for a few days only to wake up and see dozens of error messages and the bot not responding to commands.

The errors were in the form of:
```
The channel "foo" was unable to be joined. Check the channel is valid.
The channel "bar" was unable to be joined. Check the channel is valid.
...
The channel "baz" was unable to be joined. Check the channel is valid.
```
were repeated dozens of times. After those messages were done repeating, a bunch of Tracebacks started repeating as well:
```
Task exception was never retrieved
future: <Task finished name='Task-5027' coro=<WSConnection._join_future_handle() done, defined at /usr/local/lib/python3.9/dist-packages/twitchio/websocket.py:325> exception=KeyError('foo')>
Traceback (most recent call last):
  File "/usr/lib/python3.9/asyncio/tasks.py", line 492, in wait_for
    fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/twitchio/websocket.py", line 327, in _join_future_handle
    await asyncio.wait_for(fut, timeout=timeout)
  File "/usr/lib/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/twitchio/websocket.py", line 330, in _join_future_handle
    self._join_pending.pop(channel)
KeyError: 'foo'
Task exception was never retrieved
```

Should I report this as a bug, or am I missing something?

I've done some digging in websocket.py, but I couldn't find what could cause this many repeated calls.

**Comment by chillymosh** [comment, marked as answer]
I have several people testing a fix for this.
This only occurs after Twitch sends a forced RECONNECT.

In websocket.py in the `_connect` method 
Change the code to to this (It's simply adding `self._reconnect_requested = False` 

```py
        await self.authenticate(self._initial_channels)

        self._keeper = asyncio.create_task(self._keep_alive())  # Create our keep alive.
        self._reconnect_requested = False
        if not self._task_cleaner or self._task_cleaner.done():
            self._task_cleaner = asyncio.create_task(self._task_cleanup())  # Create our task cleaner.

        self._ws_ready_event.set()
```

Once confirmed everything is stable for everyone over a time period and a RECONNECT is received I'll push it.

**Reply by anakimluke** [reply]
Ok! Thanks! I'll report back if I see this behaviour once more. I also enabled debug logging to get more info should that happen again.

**Reply by 1987Krang1987** [reply]
Hi, i have the same problem now, 
was it possible to solve?

**Reply by anakimluke** [reply]
I'm still getting the same error message even after using this patch. The server has been up since my last message and it happened thrice since. It seems to have reconnected automatically this time, as I have only noticed the problem by looking at the logs, and not making a chat command directly.

Here's the debug log with a few seconds of context: https://gist.github.com/anakimluke/4de75e708fcd4122fcd59297f67877ae

**Reply by chillymosh** [reply]
You will still get the message but it reconnects now and will then stop the messages.

**Comment by chillymosh** [comment]
Just for information purposes, I have pushed the change to master.

This can be installed directly until next release.
